torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+};
+
+template <typename T>
+struct DynamicDAG {
+  ~DynamicDAG();
+
+  Vertex<T>* newVertex(T datum);
+  std::pair<vertex_set<T>,vertex_set<T>> removeVertex(Vertex<T>* v);
+
+  void addEdge(Vertex<T>* producer, Vertex<T>* consumer);
+  void removeEdge(Vertex<T>* producer, Vertex<T>* consumer);
+  bool contractEdge(Vertex<T>* producer, Vertex<T>* consumer);
+
+  bool contractionProducesCycle(Vertex<T>* producer, Vertex<T>* consumer);
+  bool dfsForward(Vertex<T>* producer, size_t upper_bound, vertex_list<T>& visited);
+  void dfsBackward(Vertex<T>* producer, size_t lower_bound, vertex_list<T>& visited);
+  void clearVisited(vertex_list<T>& visited);
+
+  void reorder(vertex_list<T>& deltaF, vertex_list<T>& deltaB);
+  void sort(vertex_list<T>& delta);
+
+  void checkInvariants();
+
+  Vertex<T>* at(size_t ord) const { return vertices_.at(ord).value(); };
+  const maybe_vertex_list<T>& maybe_vertices() const { return vertices_; };
+
+  // Use for debugging. Don't call this often.
+  size_t numVertices() const;
+
+  std::string toString();
+
+ private:
+
+  // Store vertices indexed by their topological order.
+  // If a vertex v has ord 5, then it can be found at vertices_[5].
+  // There may be gaps in vertices_; this is to enable fast deletion.
+  maybe_vertex_list<T> vertices_;
+
+  // temp buffers used in the PK algorithm.
+  // Their usage makes this data structure NOT thread safe.
+  vertex_list<T> deltaF_;  // all vertices visited by dfsForward
+  vertex_list<T> deltaB_;  // all vertices visited by dfsBackward
+};
+
+template <typename T>
+DynamicDAG<T>::~DynamicDAG() {
+  for (auto v : vertices_) {
+    if (v.has_value()) {
+      delete v.value();
+    }
+  }
+}
+
+// O(vertices_.size()). Used for testing, don't call this often.
+template <typename T>
+size_t DynamicDAG<T>::numVertices() const {
+  return std::count_if(vertices_.begin(), vertices_.end(), [](at::optional<Vertex<T>*> v) {
+      return v;
+  });
+}
+
+template <typename T>
+Vertex<T>* DynamicDAG<T>::newVertex(T datum) {
+  auto* vertex = new Vertex<T>();
+  vertex->rdata.push_back(datum);
+  vertex->visited = false;
+  vertex->ord = vertices_.size();
+  vertices_.push_back(vertex);
+  return vertex;
+}
+
+template <typename T>
+void DynamicDAG<T>::removeEdge(Vertex<T>* producer, Vertex<T>* consumer) {
+  JIT_ASSERT(producer != consumer);
+  JIT_ASSERT(producer->out_edges.count(consumer));
+  JIT_ASSERT(consumer->in_edges.count(producer));
+  producer->out_edges.erase(consumer);
+  consumer->in_edges.erase(producer);
+}
+
+template <typename T>
+void DynamicDAG<T>::checkInvariants() {
+  for (size_t ord = 0; ord < vertices_.size(); ++ord) {
+    const auto& maybe_vertex = vertices_.at(ord);
+    if (!maybe_vertex.has_value()) continue;
+    auto* vertex = maybe_vertex.value();
+
+    JIT_ASSERT(vertex->ord == ord);
+    for (auto* v : vertex->in_edges) {
+      JIT_ASSERT(v->ord < ord);
+    }
+    for (auto* v : vertex->out_edges) {
+      JIT_ASSERT(v->ord > ord);
+    }
```

Comment:
Wouldn't it be sufficient to only check `in_edges` or `out_edges`, since every edge is supposed to be in corresponding sets of its endpoints? It might be a good idea that this is the case, and then only do a single check.

Reply:
I had some memory corruption earlier from modifying an unordered_set while iterating through it that made it so that some edges were not recorded correctly so this function was helpful to debug that.

`checkInvariants` isn't called anywhere right now, I think I should rename it to `debugCheckInvariants` and use it in test_jit.cpp or something.